### ISSUE TITLE
[feat] support fp8 per-token-head kv cache with inline scale

### ISSUE THREAD

**Issue by ir1ka** [flashinfer-ai/flashinfer#3617]
vLLM supports a KV cache quantization method known as `fp8-per-token-head`, though this is currently supported only by Triton attention.

The `fp8-per-token-head` KV cache quantization generates an FP32 scale for each head, offering finer granularity compared to `fp8-per-tensor` quantization.

To facilitate KV cache management, the scales are stored inline within the KV cache and aligned to 16 bytes (ensuring compatibility with `cp.async`).

Please add support for `fp8-per-token-head` KV cache quantization to FlashInfer.

**Comment by ir1ka** [comment]
# [Feature]: FA2 FP8 per-(token, head) KV cache with inline scale (`fp8_per_token_head`)

## Problem

vLLM supports an FP8 KV-cache dynamic quantization called `fp8_per_token_head`: each (token, head) vector carries a runtime-computed float32 scale stored alongside the FP8 data. FlashInfer's FA2 kernels currently only support per-tensor-scaled FP8 KV caches and cannot directly consume a per-(token, head) quantized cache. Users (e.g. vLLM) who want to use `fp8_per_token_head` with FA2 must first dequantize back to 16-bit, losing the memory and bandwidth benefits of the FP8 cache.

## Requested outcome

All five FA2 entry points (single/batch prefill ragged/paged, single/batch decode paged, including tensor-cores decode) accept `use_inline_sf=True`: they consume an FP8 KV cache with inline per-(token, head) scales and apply the scales correctly, matching a 16-bit reference.

## Target hardware

- SM80 / SM86 / SM87 (A100, A30, RTX 30-series, Jetson Orin)
- SM89 (L4, L40S, RTX 40-series)
- SM90 (H100, H200, H20)
- Other:SM75(RTX 20 系列,如 RTX 2080 Ti)

## Inference engine

- vLLM

## Workload and configuration

- Data type and quantization: FP8 (e4m3/e5m2) KV cache with inline per-(token, head) float32 scale; 16-bit Q/O (fp16 on SM75, fp16/bf16 on SM80+)
- Batch size or request concurrency: single and batch (batch=4 in the benchmark)
- Sequence lengths or token counts: kv_len {512, 2048, 8192} in the benchmark; arbitrary in production
- Parallelism: N/A (single-GPU attention operator)
- Relevant shapes: head_dim {64, 128, 256, 192/128 asymmetric}, MHA/GQA, ragged/paged (page size 16), single/batch, causal/non-causal
- Entry points: `single_prefill_with_kv_cache`, `BatchPrefillWithRaggedKVCacheWrapper`, `BatchPrefillWithPagedKVCacheWrapper`, `single_decode_with_kv_cache`, `BatchDecodeWithPagedKVCacheWrapper` — all with `use_inline_sf=True` (tensor-cores decode included)

## Current workaround

Today FA2 cannot directly consume a `fp8_per_token_head`-quantized KV cache, so users must dequantize it back to 16-bit before calling FA2, losing the memory and bandwidth benefits of the FP8 cache.

## Impact

- Memory capacity or bandwidth

## Acceptance criteria

- Passes the 294-case correctness suite (below) across all entry points, head dims, FP8 dtypes, and Q/O dtypes.
- `inl/tensor` (inline-scale FP8 vs per-tensor FP8) stays within the measured bounds reported in the Performance section on both SM75 and SM80.

## Related work, dependencies, or suggested scope

### Design (implementation outline)

**1. Layout**

Each (token, head) K/V vector is stored as a slot:

```
[ head_dim FP8 bytes (e4m3/e5m2) ][ 4-byte float32 scale ][ pad to 16 bytes ]
```

i.e. the KV tensor's last dim = `head_dim + 16` (80 / 144 / 272 for head_dim ∈ {64, 128, 256}). The scale is the dynamic quantization factor (amax / fp8_max) of that vector, matching the semantics of vLLM's `fp8_per_token_head`.

**2. Scale staging**

All K/V scales within a tile are staged into shared memory in the same cp.async group as the K/V data (one float32 per KV row, for K and V separately), with no extra synchronization.

**3. Where the scales are applied**

| path | K scale | V scale |
|---|---|---|
| decode | before softmax, on the QK dot product, in f32 | after softmax, on the softmax output, in f32 |
| prefill | after the QK accumulation, on the S fragment, in f32 | after softmax, on the output, in the output precision (fp16/bf16) |

- decode: the K scale is applied before the softmax and the V scale after it, both in f32, to guarantee accuracy and performance.
- prefill: the K scale is applied once the QK accumulation is complete, so it is applied in f32; the V scale is applied to the output after the softmax, so it is applied in the output precision (typically fp16 or bf16).

**4. SM75 shared-memory budget**

The scale buffer consumes extra shared memory (one float32 per KV row, for K and V). On SM75 (64 KB smem per block), the FA2 prefill budget is already full in two configurations:

- head_dim=128 with CTA_TILE_Q=128 (packed_qo_len > 64): reduce the KV pipeline depth (NUM_MMA_KV) to make room for the scales;
- head_dim=256: disable the FP8→16-bit repack path (whose staging buffer is larger).

Both adjustments are gated by the compile-time template parameter `USE_INLINE_SF`: only inline-scale configs are affected, the existing per-tensor FP8 path is untouched (and serves as the performance baseline in the benchmark), and only SM75 triggers them (SM80+ has a 101376 B per-block budget, no adjustment needed).

**5. Performance**

Method: 3-way comparison (per-tensor FP8 / inline-scale FP8 / 16-bit KV), 30 iters, batch=4, page=16, num_kv_heads=8 (mha qo=8 / gqa qo=16). SM75 uses CUPTI timing with fp16 Q/O; SM80 is a CMP device that does not support CUPTI, so it uses CUDA-event timing with fp16+bf16 Q/O. The benchmark script is added by the accompanying PR.

`inl/tensor` = fp8_inline / fp8_tensor (1.0 = no regression, >1 = slower).

SM75 (RTX 2080 Ti; head_dim {64, 128, 256} × kv_len {512, 2048, 8192} × 5 entry points × {MHA, GQA}, 90 measured cells):

| head_dim | decode | prefill |
|---|---|---|
| 64 | 0.77–1.13 | 0.98–1.15 |
| 128 | 0.73–1.11 | single 0.85–0.96 · batch 1.06–1.18 |
| 256 | 0.71–1.07 | 1.13–1.57 |

- decode and head_dim=64 prefill: ≈1.0, no regression.
- head_dim=128 prefill: the CTA_TILE_Q=128 path reduces NUM_MMA_KV; batch prefill is 6–18% slower, single prefill is unchanged or faster.
- head_dim=256 prefill: REPACK is disabled; 13–57% slower (worst is batch-paged, up to 1.57x). The measured value exceeds the initial 10–30% estimate; the measured value is reported as-is.

SM80 (CMP 170HX; head_dim {128, 256} × kv_len {512, 2048} × 5 entry points × {MHA, GQA} × {fp16, bf16}, 80 measured cells; head_dim=64 is unaffected by design):

| head_dim | decode | prefill |
|---|---|---|
| 128 | 0.51–0.80 (faster) | 0.95–1.13 |
| 256 | 0.44–0.72 (faster) | 0.86–1.11 |

- prefill ≈1.0 (a few batch cells up to ~1.13x), confirming no smem adjustment is needed on SM80+ and there is no significant regression.
- decode is actually faster with inline scale (no regression).

SM89 (RTX 4070 SUPER; head_dim {64, 128, 256} × kv_len {512, 2048, 8192} × 5 entry points × {MHA, GQA} × {fp16, bf16}, 180 measured cells):

| head_dim | decode | prefill |
|---|---|---|
| 64 | 1.059–1.262 | 1.000–1.085 |
| 128 | 1.036–1.314 | 0.988–1.075 |
| 256 | 1.043–1.300 | 1.029–1.097 |

- prefill ≈1.0 (0.988–1.097), no regression, further confirming no smem adjustment is needed on SM80+.
- decode is slower than per-tensor (1.036–1.314).

all smem adjustments are gated by the compile-time template parameter `USE_INLINE_SF`, so only inline-scale configs are affected; the existing per-tensor FP8 path is untouched (and serves as the performance baseline in the benchmark). Only SM75 triggers them (SM80+ has a 101376 B per-block budget, no adjustment needed — confirmed by the measured prefill ≈1.0 on SM80 and SM89).

**6. Correctness**

294 test cases covering the 5 entry points × head_dim {64, 128, 256, 192/128 asymmetric} × {e4m3, e5m2} × {fp16, bf16} × {MHA, GQA} × {causal}, plus sliding window, logits soft cap, and CTA_TILE_Q=128 long-sequence tiling. Each case compares three outputs: the inline-scale FP8 kernel vs a 16-bit dequantized reference (cos > 0.9998, 0.998 for asymmetric 192), and vs the 16-bit original input (FP8 round-trip cos > 0.995). The test suite is added by the accompanying PR.

### Scope

FA2 backend only; no change to FA3/cuDNN paths or to the existing per-tensor FP8 path
